### DISCUSSION TITLE
Best Practice for Re-establishing Connection After Redis Datastore Down

### DISCUSSION THREAD

**Discussion by ponderMuse** [ra1u/redis-dart#95] [General]
Hi,

Trying to figure out the best way to maintain a service's connection to Redis in scenarios where a Redis datastore has been restarted.

In my current service code, I simply create one Redis client connection and use that single client throughout the service's lifetime. But, in scenarios where the Redis datastore the client is connected to is restarted, then any followup client calls will cease to work as the connection will have been broken.

Therefore, I am wondering what would be best practice when using this dart Redis library in order to automatically re-establish the client's lost connection.

The other option would be to simply create a new Redis client connection for each new incoming service call but I don't how efficient and performant this would be in comparison if the service usage is high?

Any advise is much appreciated.

**Comment by ra1u** [comment]
For second one about multiple connection is: I dont think that having multiple connection is that bad of an idea. I think that having 100 connections shout be of no problem, but going with number of connections over 1000 will require appropriate testing so that your infrastructure supports this. (use account can be limited by number of connection can single app establish).

In terms of interface you can for example extend exiting class Connection that reconnects transperently.. Once there is a disconnection, you can just open new one and resume your work over new one with same extended class. (however, you will need to notify all existing outbound packet receivers, about an error, given that they are expecting rx.)
In any case it is up to you to design usecase around apropriate error handling.

**Reply by ponderMuse** [reply]
Thanks for reply. That is pretty much what I am doing currently.

For my services, I have a wrapper class which is essentially a `RedisClient` class with a local cache of `RedisConnection`s acting as a connection Pool (with a configurable max size) which Services using the `RedisClient` class can pass in as parameter as needed.

The `RedisClient` has a call() method which picks an available `RedisConnection` to use to call the Redis datastore and indeed, if one catches an error associated with `connection loss` type problems then it will remove the `RedisConection` from the pool, attempts to establish a new connection, and re-makes the call (if connection established). But currently it does not reconnect all the `RedisConnection`s in the pool but, I could update the code to do just that.

I was more of the mind of maybe being able to set a listener on `RedisConnection`s which would get notified whenever a connection breaks instead of having to react to whenever the next `Command.call()` is made and having to re-connect there and then if the call fails because of a connection had been lost some time before. But I don't know how feasible something like this would be.
